File: ruby/organizing_policy/lib/organizing/decline_invitation.rb
require_relative 'invitation_repository'
require_relative 'invitation_status'

module Organizing
  class DeclineInvitation
    def initialize(invitation_repository)
      @invitation_repository = invitation_repository
    end

    def call(invitation_id)
      invitation = @invitation_repository.find_by_id(invitation_id)
      raise ArgumentError, "Invitation not found: #{invitation_id.value}" if invitation.nil?
      raise ArgumentError, "Invitation #{invitation_id.value} is not pending" unless invitation.status == InvitationStatus::PENDING

      @invitation_repository.save(invitation.with(status: InvitationStatus::DECLINED))
    end
  end
end

File: java/organizing-policy/src/main/java/com/hexagonalchess/organizing/DeclineInvitation.java
package com.hexagonalchess.organizing;

public class DeclineInvitation {

    private final InvitationRepository invitationRepository;

    public DeclineInvitation(InvitationRepository invitationRepository) {
        this.invitationRepository = invitationRepository;
    }

    public void execute(InvitationId invitationId) {
        var invitation = invitationRepository.findById(invitationId)
            .orElseThrow(() -> new IllegalArgumentException("Invitation not found: " + invitationId.value()));
        if (invitation.status() != InvitationStatus.PENDING) {
            throw new IllegalStateException("Invitation is not pending");
        }
        invitationRepository.save(new GameInvitation(
            invitation.id(), invitation.invitingUserId(), invitation.invitedUserId(), InvitationStatus.DECLINED
        ));
    }
}
